SetGame/SetGame/SetGameModel.swift

```diff
@@ -0,0 +1,274 @@
+    //check if selected cards are set -> replace them, else add three cards
+    mutating func addThreeCardsButtonPressed() {
+        var indexes = self.clearSelectedSet()
+        self.addCardsToBoard(at: indexes)
+    }
+    
+    //if choosen cards are set - clear selected, change state to chosen and return indexes
+    private mutating func clearSelectedSet() -> [Int]? {
+        var res: [Int] = []
+        var toRemove: [Card] = []
+        if self.choosenCardsState == CardState.match {
+            //find indexes
+            for card in self.selectedCards {
+                res.append(self.cardOnBoard.firstIndex(of: card)!)
+                toRemove.append(card)
+            }
+            //remove from board
+            for card in toRemove {
+                self.cardOnBoard.remove(element: card)
+            }
+            self.selectedCards = [Card]()
+            self.choosenCardsState = .chosen
+            return res.sorted()
+        }
+        return nil
+    }
+    
+    func isSet() -> Bool {
+        return compareCardsValues(by: {
+            let b1 = ($0 == $1 && $1 == $2)
+            let b2 = ($0 != $1 && $1 != $2 && $2 != $0)
+            return b1 || b2 })
+//        return compareCardsValues(by: { _,_,_ in return true })
+    }
+    
+    //compare all selected cards attributes (color, filling, shape, shapeCount) by boolean comparison function
+    private func compareCardsValues(by comparison: (CardProperty, CardProperty, CardProperty) -> Bool) -> Bool {
+        return comparison(self.selectedCards[0].color, self.selectedCards[1].color, self.selectedCards[2].color) &&
+            comparison(self.selectedCards[0].filling, self.selectedCards[1].filling, self.selectedCards[2].filling) &&
+            comparison(self.selectedCards[0].shape, self.selectedCards[1].shape, self.selectedCards[2].shape) &&
+            comparison(self.selectedCards[0].shapeCount, self.selectedCards[1].shapeCount, self.selectedCards[2].shapeCount)
+    }
+    
+    //return 3 indexes of existing set, or nil if not exist
+    mutating func getHint() -> (Int, Int, Int)? {
+        //if selected card already a match - first replace selected cards
+        if self.choosenCardsState == .match {
+            let indexes = self.clearSelectedSet()
+            self.addCardsToBoard(at: indexes)
+        }
+        //reduce score by one
+        self.score -= 1
+        return self.findSet()
+    }
+    
+    //return 3 indexes of existing set, or nil if not exist
+    func findSet() -> (Int, Int, Int)? {
+        for var i in 0..<cardOnBoard.count {
+            for var j in i+1..<cardOnBoard.count {
+                if let k = thirdCardForSetExist(index1: i, index2: j) {
+                    return (i, j, k)
+                }
+            }
+        }
+        return nil
+    }
+    
+    //check if board have third card to complete the given 2 cards to a set, and return the index
+    func thirdCardForSetExist(index1: Int, index2: Int) -> Int? {
```

Comment:
I love the implementation of this method, the use of `thirdPropertyForSet` is brilliant, really nice :)